File: backend/app/schemas/fee.py
from typing import Optional, List
from pydantic import BaseModel
from datetime import datetime
from enum import Enum

class FeeInterval(str, Enum):
    MONTHLY = "MONTHLY"
    YEARLY = "YEARLY"
    ONE_TIME = "ONE_TIME"

class PaymentStatus(str, Enum):
    PENDING = "PENDING"
    PAID = "PAID"
    FAILED = "FAILED"

# Fee Schemas
class MembershipFeeBase(BaseModel):
    name: str
    amount: float
    interval: Optional[FeeInterval] = FeeInterval.YEARLY
    description: Optional[str] = None
    is_active: bool = True

class MembershipFeeCreate(MembershipFeeBase):
    pass

class MembershipFee(MembershipFeeBase):
    id: int
    class Config:
        from_attributes = True

# Payment Schemas
class PaymentBase(BaseModel):
    amount: float
    status: PaymentStatus = PaymentStatus.PENDING
    transaction_id: Optional[str] = None

class PaymentCreate(BaseModel):
    fee_id: int

class Payment(PaymentBase):
    id: int
    user_id: int
    fee_id: int
    payment_date: Optional[datetime] = None
    created_at: datetime
    fee: Optional[MembershipFee] = None
    
    class Config:
        from_attributes = True

File: frontend/src/services/fee.ts
import { fetchJson } from "@/lib/api";

export interface MembershipFee {
    id: number;
    name: string;
    amount: number;
    interval: "MONTHLY" | "YEARLY" | "ONE_TIME";
    description?: string;
    is_active: boolean;
}

export interface CreateFeeDTO {
    name: string;
    amount: number;
    interval: "MONTHLY" | "YEARLY" | "ONE_TIME";
    description?: string;
}

export interface Payment {
    id: number;
    user_id: number;
    fee_id: number;
    amount: number;
    status: "PENDING" | "PAID" | "FAILED";
    transaction_id?: string;
    payment_date?: string;
    created_at: string;
    fee?: MembershipFee;
}

export const feeService = {
    async getFees(): Promise<MembershipFee[]> {
        return fetchJson<MembershipFee[]>("/api/v1/fees/fees");
    },

    async createFee(data: CreateFeeDTO): Promise<MembershipFee> {
        return fetchJson<MembershipFee>("/api/v1/fees/fees", {
            method: "POST",
            body: JSON.stringify(data),
        });
    },

    async makePayment(feeId: number): Promise<Payment> {
        return fetchJson<Payment>("/api/v1/fees/payments", {
            method: "POST",
            body: JSON.stringify({ fee_id: feeId }),
        });
    },

    async getMyPayments(): Promise<Payment[]> {
        return fetchJson<Payment[]>("/api/v1/fees/payments/me");
    },

    async getAllPayments(): Promise<Payment[]> {
        return fetchJson<Payment[]>("/api/v1/fees/payments");
    }
};
